### DISCUSSION TITLE
Web GUI for PicoW

### DISCUSSION THREAD

**Discussion by Deftaudio** [ZuluSCSI/ZuluSCSI-Pico-OSHW#4] [Ideas]
Hi Alex and the team. 
I understand the primary motivation for PicoW is a Daynaport emulation, but would you consider adding a Web interface similar to PiSCSI for selecting images? It would be very convenient to change them, even if a unit restart required.

**Comment by Deftaudio** [comment]
Just check on this, you think it could be possible to implement?

**Comment by Deftaudio** [comment]
Hi [aperezbios](https://github.com/aperezbios)
would you, please, let me know if it's possible?

**Comment by aperezbios** [comment]
@Deftaudio it's not practical to implement this due to memory constraints. There is simply no room for a TCP stack and web server.

**Comment by Deftaudio** [comment]
I'm not sure, I understand, @aperezbios . You have room for TCP stack since you can run DaynaPORT emulation and connect to wifi. if you don't have room for a web server then telnet / ssh can work too. Is it feasible?

**Reply by ZigZagJoe** [reply]
The TCP stack is handled on the host computer. The SCSI emulator just transfers frames, nothing more, so it's much lighter weight but has no awareness of any higher protocols such as TCP/IP.

**Comment by aperezbios** [comment]
If you don't believe me, feel free to do the work yourself. This sort of second-guessing is really unhelpful and comes across as ignorantly disrespectful. If there's not a straightforward path to a given feature request, generally it won't get implemented. These are resource-constrained microcontrollers we're talking about here, not Linux SBCs.

**Comment by Deftaudio** [comment]
I'm sorry @aperezbios  I didn't mean to be disrespectful and not sure why you'd think so. I think I deserve a little bit more explanation, more than one line after a year of pinging you and sending direct message, and by the way buying your products for several thousands, as you know. 
So, please, spend a minute of your time explaining this to me. 

@ZigZagJoe, correct it's sent as SCSI frames, but isn't PicoW connects to WiFi and repacks those into ethernet frames? So, you're saying there is no room to parse TCP/IP? Would extra PSRAM help to manage it?

**Reply by ZigZagJoe** [reply]
They are sent/received as ethernet frames first which are then encapsulated via vendor-specific SCSI commands, and unpacked and presented as frames on the host side. So there's no higher understanding on the Pico, beyond the mechanics of transferring the raw frames over SCSI.

It wouldn't be conceptually impossible to support higher protocol layers but it'd require RAM and CPU time to support. It would be more practical to have magic frames which it could recognize and use for management commands, but that'd require a smart client application on the device accessing the zuluscsi. Given that, practically, I think it would make much more sense to use bluetooth.

**Reply by aperezbios** [reply]
@Deftaudio I don't think it was your intent to be, just explaining that it's how it came across to me. You still have to have somewhere to attach extra PSRAM to. There are _no free available I/O's for this_ on any RP2040/Pico-based ZuluSCSI product. The RP2040 is a great microcontroller, but it has very limited I/O's, which is why our newest ZuluSCSI family member uses the ZuluSCSI RP2350B, which has additional I/O's over the RP2040/RP2350A. There is insufficient memory and available I/O to accomplish this task with the given hardware.

**Reply by Deftaudio** [reply]
Thank you both for the reply. I saw the announcement of RP2350B based product, impressive performance boost. 
Do you think it might be possible that OSHW design to move to Pico2 in future? and so then, Pico2 W with extra PSRAM might be used? Such as "Pimoroni Pico Plus 2 W" or similar clones.
 
Doesn't need to be OSHW, by the way, your commercial RP2350B based product with added WiFi or BT capability to reconfigure the INI file. No data transfers, just image selection.

**Reply by aperezbios** [reply]
The existing Pico-based designs can already be used with Pico 2 and 2W's. The only difference is the firmware, which is already available. While this largely "solves" the SRAM availability problem, it's far from a cure-all, because you still have a finite amount of compute power. The primary responsibility of the ZuluSCSI firmware is to service I/O requests in as short a time as possible. Mixing this with non-realtime workloads such as a web server and TCP stack is still not a great engineering decision, but if you'd like to work on it, more power to you :)

**Reply by Deftaudio** [reply]
Totally get that any service will be on a way to impact I/O timings. Even one can argue that you only select images when you don't do any SCSI I/O, you still have to run the stack. 

I'm so badly missing a way to select images without taking out SD Card. 
Maybe you could consider then the optional OLED Screen with Encoder to select those?

**Reply by ZigZagJoe** [reply]
I implemented USB Mass storage in zuluscsi for this reason - you can connect a USB cable (while the host is off!) to a computer and the SD card will show up as a storage device. Slow, not suited for transferring files but does work for zuluscsi.ini edits and moving files in and out of a "disabled" folder.

**Reply by aperezbios** [reply]
@Deftaudio for specifics of how to use the functionality @ZigZagJoe mentioned, see the documentation at https://github.com/ZuluSCSI/ZuluSCSI-firmware/wiki/USB-Mass-Storage

**Comment by Deftaudio** [comment]
Thank you, as I'm aware and familiar of that recent feature. it's great to have it but this doesn't solve my problem unfortunately. 
My use case is changing CD images, while the system is running. Powering it off and connecting the cable (and then disconnecting) is not any better than taking the SD card out. Really need a PiSCSI approach here to have an option to select the image while running (with the proper SCSI eject handling) without quirks of PiSCSI :) 
Thought first having a Web page for that, now reduced expectations to the local control :)

**Comment by aperezbios** [comment]
@Deftaudio which operating system are you looking to accomplish this under? Without knowing that, I can't recommend the best path forward.

**Reply by Deftaudio** [reply]
Hardware samplers - Akai, EMU, Kurzweil, etc. PiSCSI does it fine, I may end up having ZuluSCSI + PiSCSI together. One to boot HDD immediately at power on and PiSCSI to manage CDs only.

**Reply by aperezbios** [reply]
You  should always use the tools that work best for your workflow. PiSCSI has functionality that ZuluSCSI doesn't have, and vice versa. I see zero problem with using two separate solutions to achieve your goals.
